src/accelerate/utils/fsdp_utils.py

```diff
@@ -207,3 +211,66 @@ def load_fsdp_optimizer(fsdp_plugin, accelerator, optimizer, model, input_dir, o
             logger.info(f"Optimizer loaded from {ckpt_dir}")
         flattened_osd = FSDP.optim_state_dict_to_load(model=model, optim=optimizer, optim_state_dict=optim_state)
         optimizer.load_state_dict(flattened_osd)
+
+
+def _distributed_checkpoint_to_merged_weights(checkpoint_dir: str, save_path: str, use_safetensors: bool = True):
+    """
+    Passthrough to `torch.distributed.checkpoint.format_utils.dcp_to_torch_save`
+
+    Will save to `{save_path}/merged.bin`.
+    """
+    state_dict = {}
+    save_path = Path(save_path)
+    dist_cp_format_utils._load_state_dict(
+        state_dict,
+        storage_reader=dist_cp.FileSystemReader(checkpoint_dir),
+        planner=dist_cp_format_utils._EmptyStateDictLoadPlanner(),
+        no_dist=True,
+    )
+    # if we just have a folder, use merged
+    if save_path.suffix == "":
+        fname = "merged"
+        if use_safetensors:
+            fname += ".safetensors"
+        else:
+            fname += ".bin"
+        save_path = save_path / fname
+
+    # To handle if state is a dict like {model: {...}}
+    if len(state_dict.keys()) == 1:
+        state_dict = state_dict[list(state_dict)[0]]
+    save(state_dict, save_path, safe_serialization=use_safetensors)
+    return save_path
+
+
+def merge_fsdp_weights(
+    checkpoint_dir: str, output_path: str, use_safetensors: bool = True, remove_checkpoint_dir: bool = False
+):
+    """
+    Merge the weights from multiple FSDP checkpoints into a single combined checkpoint. Should be used if
+    `SHARDED_STATE_DICT` was used for the model. Weights will be saved to `{output_path}/merged.pth`.
+
+    Note: this is a CPU-bound process.
+
+    Args:
+        checkpoint_dir (`str`):
+            The directory containing the FSDP checkpoints (can be either the model or optimizer).
```

Comment:
what happens if we only have the optimizer ?

Reply:
It can be one or the other :)